### ISSUE TITLE
Documentation for creating own components needed (e.g. model, connectivity, synapse dynamics, activation funciton etc.)

### ISSUE THREAD

**Issue by AaronSpieler** [norse/norse#288]
Hello

so I was trying to implement my own neuron model, synapse dynamic, connectivity etc., however,  I did not find any guide/notebook on how to do that. As far as I can see the tutorials and documentation are only about how to use the existing parts. 

And as I try to figure out the structure of the library its quite overwhelming to just write a slightly different neuron model (even though the code looks clean). Especially because just looking at LIF model as an example on how to do it there are so many configuration that its hard to see what are the minimal required definitons.

I would suggest adding a tutorial on:
- intro in general: how is the general structure of the library (e.g. components) and how they interact
- how you would implement your minimal working neuron model (e.g. basic LIF) from scratch in this library
- how synapse dynamics would be implemented (e.g. STDP)
- how specific connectivity would be implemented (e.g. sparse)
- how to implement own activation function

Basically something analougous to "Usage Docs", but for creating your own components e.g. "Creation Docs".
This would help immensely in being able to use your library for own research (beyond just using existing components).

This is related to:
- https://github.com/norse/norse/issues/94
- https://github.com/norse/norse/issues/224
- https://github.com/norse/norse/issues/87
- https://github.com/norse/norse/issues/88
- https://github.com/norse/norse/issues/278
- https://colab.research.google.com/github/norse/notebooks/blob/master/stp_example.ipynb (seems to be a bit incomplete?)

Best,
Aaron

P.s. if I'm mistaken and this already exists out there I would appreciate a link

**Comment by cpehle** [comment]
This is a point well taken, we should definitely look into improving documentation related to this. Right now there is fairly little in the way of actual "framework" setup. However I can maybe document the current way of adding a neuron model in more detail. In the meantime, you might also take a look at one of the pull requests that implemented a new neuron model. You are also welcome to join our discord (https://discord.gg/MASdh5m) and we are planning to do a CodeJam fairly soon, where such developer focussed issues can be discussed directly and synchronously :).

**Comment by Jegp** [comment]
Excellent idea! I can only agree with @cpehle.

If you can document adding neuron models, @cpehle, I can perhaps try to document the library structure slightly.

While you, @AaronSpieler, wait for us lazy academics to get moving, I can inform you that our structure closely resemble that of PyTorch in that we separate **basic functionality** (`norse.torch.functional`) from **neural network modules** (`norse.torch.module`). The added complexity is the notion of time, where we distinguish between `Cell` and non-`Cell` modules ([described tangentially here](https://norse.github.io/norse/pages/working.html#neuron-dynamics-and-time)). So, the simplest way to add a neuron model would be to 
1. Define the functional part
2. Wrap it in a nice `Cell` module (without time)

Just note that one of the reasons this hasn't been well documented is that there are some changes on the horizon (#242). So, unfortunately, it may take a bit of time for us to arrive at a "finished" model-building guide. Any contribution/input/participation is naturally appreciated and encouraged.

**Comment by Jegp** [comment]
I started addressing it slowly in PR #290. Suggestions and contributions are most welcome. I feel it's going to be hard to arrive at anything that feels "done" anytime soon, so I would suggest we add sections slowly one by one. This particular PR is meant to address the architecture and implementation detail.

**Comment by Jegp** [comment]
Btw, if you have any further ideas/suggestions/code to improve on this, @AaronSpieler, we're more than interested to keep talking :-)

**Comment by AaronSpieler** [comment]
Thx for the quick response!

The documentation is now very helpful to get a first overview.
Some details are still missing, esp. connectivity, but I guess you will add this as the library matures.

I will try to work on my stuff and see whether I bump into problems.

Thx for your work on this library!
